mirrord-agent/src/main.rs

```diff
@@ -160,70 +161,99 @@ async fn handle_peer_messages(
             let peer = state.peers.get(&peer_message.peer_id).unwrap();
             peer.channel.send(DaemonMessage::Pong).await?;
         }
-        ClientMessage::FileRequest(file_request) => {
-            debug!(
-                "ClientMessage::FileRequest peer id {:?} requesting {:?}",
-                peer_message.peer_id, file_request
-            );
-
-            file_request_tx
-                .send((peer_message.peer_id, file_request))
-                .await?;
+        ClientMessage::FileRequest(_) => {
+            // handled by the peer..
         }
     }
 
     Ok(())
 }
 
-async fn peer_handler(
-    mut daemon_messages_rx: mpsc::Receiver<DaemonMessage>,
-    peer_messages_tx: mpsc::Sender<PeerMessage>,
-    stream: TcpStream,
-    peer_id: PeerID,
-) -> Result<(), AgentError> {
-    let mut daemon_stream = actix_codec::Framed::new(stream, DaemonCodec::new());
+struct PeerHandler {
+    pub sender: Sender<PeerMessage>,
+    id: PeerID,
+    file_manager: FileManager,
+    stream: Framed<TcpStream, DaemonCodec>,
+    receiver: Receiver<DaemonMessage>,
+}
 
-    loop {
-        select! {
-            message = daemon_stream.next() => {
-                debug!("peer_handler -> daemon_stream.next received a message {:?}", message);
-
-                match message {
-                    Some(message) => {
-                        let message = PeerMessage {
-                            client_message: message?,
-                            peer_id
-                        };
-
-                        debug!("peer_handler -> client sent message {:?}", message);
-                        peer_messages_tx.send(message).await?;
-                    }
-                    None => break
+impl PeerHandler {
+    /// A loop that handles peer connection and state. Brekas upon receiver/sender drop.
+    pub async fn start(
+        sender: Sender<PeerMessage>,
+        id: PeerID,
+        stream: TcpStream,
+        receiver: Receiver<DaemonMessage>,
+        pid: Option<u64>,
+    ) -> Result<(), AgentError> {
+        let file_manager = FileManager::new(pid);
+        let stream = actix_codec::Framed::new(stream, DaemonCodec::new());
+        let mut peer_handler = PeerHandler {
+            sender,
+            id,
+            file_manager,
+            stream,
+            receiver,
+        };
+        peer_handler.handle_loop().await?;
+
+        peer_handler
+            .sender
+            .send(PeerMessage {
+                client_message: ClientMessage::Close,
```

Comment:
Is this the agent sending the layer a layer->agent message?

Reply:
erm it's a bit off but yeah :F I think we can fix it after I refactor the sniffer part.